presto-main/src/main/java/com/facebook/presto/execution/scheduler/group/DynamicLifespanScheduler.java

```diff
@@ -107,39 +132,62 @@ public void onLifespanExecutionFinished(Iterable<Lifespan> newlyCompletelyExecut
         synchronized (this) {
             for (Lifespan newlyCompletelyExecutedDriverGroup : newlyCompletelyExecutedDriverGroups) {
                 checkArgument(!newlyCompletelyExecutedDriverGroup.isTaskWide());
-                recentlyCompletelyExecutedDriverGroups.add(newlyCompletelyExecutedDriverGroup);
+                availableTasks.enqueue(taskByDriverGroup[newlyCompletelyExecutedDriverGroup.getId()]);
                 totalLifespanExecutionFinished++;
+                int driverGroupId = newlyCompletelyExecutedDriverGroup.getId();
+                runningDriverGroupIdsByTask[taskByDriverGroup[driverGroupId]].remove(driverGroupId);
             }
             newDriverGroupReady = this.newDriverGroupReady;
         }
         newDriverGroupReady.set(null);
     }
 
+    @Override
+    public boolean retryFailedTask(int taskId, List<SourceScheduler> sourceSchedulers)
```

Comment:
`retryFailedTask` looks ambiguous. What it did is:
- cleanup the failed task's lifespan
- enqueue driver groups originally assigned to this task. 

There is no retry happen (yet).

-----------------------------------------------

Also, I just realized this makes `LifespanScheduler` know task id, which is clearly not ideal. Although I don't have a better suggestion yet and I have to admit all these `LifespanScheduler`, `StageScheduler` and `SourceScheduler` are already tightly coupled and is a total mess.

I will have to think about how to clean up this mess...

Reply:
Per offline discussion, I changed the signature of this method to
```
void onTaskFailed(int task, List<SourceScheduler> sourceSchedulers)
        throws TooManyTaskFailedException;
```

`boolean` is a confusing return type for any of the proposed name (even with the original `retry` name), so instead we used exception to indicate that retry does not happen as expected.